File: app/src/main/java/wxm/KeepAccount/ui/data/show/note/ListView/ActionHelper.java
package wxm.KeepAccount.ui.data.show.note.ListView;

import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import butterknife.BindDrawable;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import wxm.KeepAccount.R;
import wxm.androidutil.util.UtilFun;

/**
 * helper for action layout
 * Created by ookoo on 2018/3/26.
 */
public abstract class ActionHelper {
    @BindView(R.id.iv_show_tag)
    ImageView mIVHideShow;

    @BindView(R.id.rl_hide_show)
    RelativeLayout mRLHideShow;

    @BindDrawable(R.drawable.ic_to_left)
    Drawable mDAExpand;

    @BindDrawable(R.drawable.ic_to_right)
    Drawable mDAHide;

    @BindView(R.id.rl_acts)
    RelativeLayout mRLActions;

    @BindView(R.id.rl_action)
    RelativeLayout mRLAction;

    @BindView(R.id.rl_lv_note)
    RelativeLayout mRLLVNote;

    private boolean  mIsShow;

    ActionHelper()    {
        mIsShow = true;
    }

    /**
     * bind layout for ButterKnife
     * @param holder        father view
     */
    public void bind(View holder)   {
        ButterKnife.bind(this, holder);
    }

    /**
     * switch bottom action bar show/hide
     * @param v         param
     */
    @OnClick(R.id.rl_hide_show)
    public void switchActionHideShow(View v) {
        setActsVisibility(!mIsShow);
        mIsShow = !mIsShow;
    }

    /**
     * use this to init self actions
     */
    protected abstract void initActs();

    /**
     * invoke this before use
     */
    protected void init()   {
        setActsVisibility(mIsShow);

        if(mIsShow) {
            initActs();
        }
    }

    /**
     * set visibility for self
     * @param bshow     show UI if true
     */
    private void setActsVisibility(boolean bshow)   {
        ViewGroup.LayoutParams rp = mRLActions.getLayoutParams();

        mIVHideShow.setImageDrawable(bshow ? mDAHide : mDAExpand);
        mRLHideShow.getBackground().setAlpha(bshow ? 255 : 40);
        mRLAction.getBackground().setAlpha(bshow ? 255 : 0);

        rp.width = bshow ? ViewGroup.LayoutParams.MATCH_PARENT : 0;
        mRLActions.setLayoutParams(rp);

        RelativeLayout.LayoutParams rp_lv = UtilFun.cast_t(mRLLVNote.getLayoutParams());
        if (bshow)
            rp_lv.addRule(RelativeLayout.ABOVE, R.id.rl_action);
        else
            rp_lv.removeRule(RelativeLayout.ABOVE);
        mRLLVNote.setLayoutParams(rp_lv);
    }
}

File: app/src/main/java/wxm/KeepAccount/ui/data/show/note/ListView/ActionHelper.kt
package wxm.KeepAccount.ui.data.show.note.ListView

import android.graphics.drawable.Drawable
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import wxm.KeepAccount.R
import wxm.KeepAccount.improve.let1
import wxm.androidutil.app.AppBase
import wxm.androidutil.ui.view.EventHelper

/**
 * helper for action layout
 * Created by WangXM on 2018/3/26.
 */
abstract class ActionHelper internal constructor() {
    private lateinit var mIVHideShow: ImageView
    private lateinit var mRLHideShow: RelativeLayout
    private lateinit var mRLActions: RelativeLayout
    private lateinit var mRLAction: RelativeLayout
    private lateinit var mRLLVNote: RelativeLayout

    private val mDAExpand: Drawable = AppBase.getDrawable(R.drawable.ic_to_left)
    private val mDAHide: Drawable = AppBase.getDrawable(R.drawable.ic_to_right)

    private var mIsShow: Boolean = false

    init {
        mIsShow = true
    }

    /**
     * bind layout for ButterKnife
     * @param holder        father view
     */
    fun bind(holder: View) {
        mIVHideShow = holder.findViewById(R.id.iv_show_tag)
        mRLHideShow = holder.findViewById(R.id.rl_hide_show)
        mRLActions = holder.findViewById(R.id.rl_acts)
        mRLLVNote = holder.findViewById(R.id.rl_lv_note)
        mRLAction = holder.findViewById(R.id.rl_action)

        EventHelper.setOnClickOperator(holder,
                intArrayOf(R.id.rl_hide_show),
                { _ ->
                    setActsVisibility(!mIsShow)
                    mIsShow = !mIsShow
                })

        setActsVisibility(mIsShow)
        if (mIsShow) {
            initActs(holder)
        }
    }


    /**
     * use this to init self actions
     */
    protected abstract fun initActs(parentView: View)


    /**
     * if [bShow] is true set view visibility
     */
    private fun setActsVisibility(bShow: Boolean) {
        val rp = mRLActions.layoutParams

        mIVHideShow.setImageDrawable(if (bShow) mDAHide else mDAExpand)
        mRLHideShow.background.alpha = if (bShow) 255 else 40
        mRLAction.background.alpha = if (bShow) 255 else 0

        rp.width = if (bShow) ViewGroup.LayoutParams.MATCH_PARENT else 0
        mRLActions.layoutParams = rp

        (mRLLVNote.layoutParams as RelativeLayout.LayoutParams).let1 {
            if (bShow)
                it.addRule(RelativeLayout.ABOVE, R.id.rl_action)
            else
                it.removeRule(RelativeLayout.ABOVE)
            mRLLVNote.layoutParams = it
        }
    }
}
